### PR TITLE
Update pod readiness condition

### PR DESCRIPTION
**Pull request by solmonk**
**What type of PR is this?**

feature

**Which issue does this PR fix**:

#560

**What does this PR do / Why do we need it**:

Please refer to the issue description of #560 for the background.

* Model build: now ready and non-ready targets are both registered (unless they are terminating)
* Target synthesizer: updates pod readiness condition. Only operates when the pod has the readiness gate. It is mostly injected by webhook, but people may add gate by themselves (that's how I tested)
  * Might be slightly off placed but couldn't find better place to fit in.
  * Status logic is mostly in the synthesizer as well, I tried to comment as much as possible.
  * I put status update logic in the PostSynthesize step, as otherwise health check status can be misleading w/ unused status, etc. 
* Target manager: stale target identification logic slightly improved

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:

**Testing done on this change**:

Tested in a simple scenario where I have 2 pods in the deployment, and adding pods. (w/ readiness gate)

* Confirmed pods are starting with non-ready state
* Confirmed targets are added while pods are non ready
* Confirmed pods are becoming ready on the next reconciliation loop

Will update after further test.

**Automation added to e2e**:
e2e test cases coming soon

**Will this PR introduce any new dependencies?**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by liwenwu-amazon** [comment]
Can you add some detail design note on how often do controller reconcile on all lattice TG targets? 
e.g. 
Is it on the timer-interval?

Does it pull lattice health state for all targets/pods?

**Comment by solmonk** [comment]
No - the controller only listens on endpoint event, and pulls lattice health state if any of the endpoint is not ready. This means the controller does not go in opposite direction (making ready pods non-ready)

The purpose of readiness is about "disabling traffic while keeping pods alive" which is already handled by VPC Lattice HC, so I believe this should not block any use cases, although we can support this in the future.

**Comment by liwenwu-amazon** [comment]
> No - the controller only listens on endpoint event, and pulls lattice health state if any of the endpoint is not ready. This means the controller does not go in opposite direction (making ready pods non-ready)
> 
What happens if targets HC are not ready ?  Will controller retry later or just busy waiting

**Comment by liwenwu-amazon** [comment]
To make the PR smaller, Is it possible to have a separate CR just on endpoint slice change?

**Comment by solmonk** [comment]
controller just goes to another reconcile loop, not busy waiting.
actually a separated PR was already up if you're interested - https://github.com/aws/aws-application-networking-k8s/pull/604
If this gets merged first I can rebase this PR to reduce the diff.

**Review by erikfuller — COMMENTED** [review]
I'm still working through some of the tests but wanted to publish the comments I had in the meantime. Mostly I just have lots of questions.

File: pkg/deploy/lattice/targets_synthesizer.go [hunk 1]
```diff
@@ -58,6 +73,113 @@ func (t *targetsSynthesizer) Synthesize(ctx context.Context) error {
...
+		if err != nil {
+			return err
+		}
+
+		identifier := model.TgNamePrefix(tg.Spec)
+		if tg.Status != nil && tg.Status.Id != "" {
+			identifier = tg.Status.Id
+		}
+
+		latticeTargets, err := t.targetsManager.List(ctx, tg)
```
[the remark this answers is not in the document]
Comment (solmonk) [reply pkg/deploy/lattice/targets_synthesizer.go:95]: This actually makes sense. Not sure if I can fit it in this PR, but will give it a try.

File: pkg/gateway/model_build_targets.go [hunk 2]
```diff
@@ -169,28 +171,35 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
...
+						if aws.BoolValue(ep.Conditions.Terminating) {
+							continue
+						}
+						target := model.Target{
+							TargetIP: address,
+							Port:     int64(aws.Int32Value(port.Port)),
+							Ready:    aws.BoolValue(ep.Conditions.Ready),
+						}
+						if ep.TargetRef != nil && ep.TargetRef.Kind == "Pod" {
+							target.TargetRef = types.NamespacedName{Namespace: ep.TargetRef.Namespace, Name: ep.TargetRef.Name}
```
[the remark this answers is not in the document]
Comment (erikfuller) [reply pkg/gateway/model_build_targets.go:199]: ah of course. Thanks for confirming

File: pkg/deploy/lattice/targets_synthesizer_test.go [hunk 3]
```diff
@@ -61,3 +68,190 @@ func Test_SynthesizeTargets(t *testing.T) {
...
+			lattice:        newLatticeTarget("10.10.1.1", 8675, vpclattice.TargetStatusUnhealthy),
+			pod:            newPod("ns", "pod1", true, true),
+			expectedStatus: corev1.ConditionTrue,
+			expectedReason: ReadinessReasonHealthy,
+			requeue:        false,
+		},
+		{
+			name:           "Unused pods keep condition",
+			model:          target,
+			lattice:        newLatticeTarget("10.10.1.1", 8675, vpclattice.TargetStatusUnused),
```
Comment (erikfuller) [inline pkg/deploy/lattice/targets_synthesizer_test.go:200]: Should this include a test for ```Draining``` just for completeness?

**Review by erikfuller — COMMENTED** [review]
Looked through the tests. No major concerns. If we can figure out the multiple target group case we're probably good to go.

**Review by erikfuller — APPROVED** [review]
Thanks for putting this together and for all your efforts and leadership in this project. I think this PR gives us a good starting point, though I agree we may have to introduce multiple readiness gates as is done in load balancer controller.

While the logic all looks correct, we should still verify this with a zero-downtime rolling update, as that is the ultimate use case here. I'll see if someone can follow up with this as part of pre-release testing.

**Review by zijun726911 — APPROVED** [review]

**Pull request merged**